src/coreclr/src/jit/importer.cpp

```diff
@@ -4015,6 +4015,75 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 break;
             }
 
+            case NI_System_Type_get_IsClass:
+            case NI_System_Type_get_IsPrimitive:
+            case NI_System_Type_get_IsValueType:
+            {
+                // Optimize things like
+                //
+                //   ldtoken [Type]
+                //   call Type.GetTypeFromHandle (which is replaced with CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE)
+                //   call Type.IsValueType (also IsClass and IsPrimitive)
+                //
+                // to `true` or `false`
+                // e.g. `typeof(int).IsValueType` => `true`
+                if (impStackTop().val->IsCall())
+                {
+                    GenTreeCall* call = impStackTop().val->AsCall();
+                    if (call->gtCallMethHnd == eeFindHelper(CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE))
+                    {
+                        GenTree* arg = call->Args().begin()->GetNode();
+                        if (!arg->IsIntegralConst())
+                        {
+                            // e.g. GT_RUNTIMELOOKUP
+                            break;
+                        }
+                        GenTreeIntCon* handle    = arg->AsIntCon();
+                        auto           hClass    = reinterpret_cast<CORINFO_CLASS_HANDLE>(handle->IconValue());
+                        DWORD          classAttr = info.compCompHnd->getClassAttribs(hClass);
+                        if (classAttr & CORINFO_FLG_SHAREDINST)
```

Comment:
This is check is too strict. We really just need to test that the type is not `Canon`. E.g. if it is `KeyValuePair<Canon, Canon>`, we can let it through.

Reply:
Not sure how to properly do that but I've found [canInlineTypeCheckWithObjectVTable](https://github.com/dotnet/runtime/blob/bc47321f0c1922408fa07e1953e5b9b5735cf7d2/src/coreclr/src/vm/jitinterface.cpp#L3855) so probably can serve as a `Is __Canon` check (used it in my latest commit)

Reply:
Maybe make `__Canon` a well known class too? I have wanted to test for it explicitly in the past.

Reply:
Added to the list of well-knowns